Add frontend arithmetic ops and shape property

jacobhinkle opened the pull request:
This adds convenient ops to the `Tensor` and `Scalar` types in the frontend. Each of those types was extended to include a pointer to the `FusionDefinition` that defined it, which it uses to translate operations into calls to `fd.ops.*`. The ops let us write stuff like `tv3 = tv1 + tv2 / tv0 + c0`.

This is related to the third task in https://github.com/csarofeen/pytorch/pull/2449. Once that PR is merged, we can include numeric literals in these expressions as well, such as `tv4 = 1 + tv2 ** 2`.

Note that extracting dtype from the frontend types is not included in this PR. Determining dtype for frontend Tensors requires accurately mimicking the type promotion machinery in `arith.cpp`. I'll approach that in a separate pull request.

jacobhinkle committed:
Add fd member to Tensor and Scalar

jacobhinkle committed:
Add .shape property to Tensor

This also includes a fix for the printing of the TensorSizesRecord, so
that it always prints as a tuple assignment. Without that change, the
test was failing when the printed python definition was evaluated.

jacobhinkle committed:
Add unary ops __neg__ and __abs__ to Tensor/Scalar

Since we don't have an inv() op, we can't add __inv__ (i.e. ~x). So I
think these are the only unary ops we have. I will add binary ops next.

jacobhinkle committed:
Add special Python binary operators to Tensor and Scalar.

See the note about `__div__` not being implemented. We could implement
it if we need to, but since we don't even have the op currently, I'm
ignoring it for now.

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:
```diff
@@ -77,6 +77,26 @@ void initNvFuserPythonBindings(PyObject* module) {
+  tensor_class.def("__add__", [](nvfuser::Tensor& self) {});
+
+  tensor_class.def(
+      "get_fusion_definition", [](nvfuser::Tensor& self) { return self.fd; });
```
Looks like an API that's simply used for debugging, maybe `_get_fusion_definition` instead.

jacobhinkle replied:
Yeah we can hide it. Although it might also be useful in user code, for example if you wanted to write a library of functions to use when building fusions.

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:
```diff
@@ -77,6 +77,26 @@ void initNvFuserPythonBindings(PyObject* module) {
+  tensor_class.def(
+      "get_fusion_definition", [](nvfuser::Tensor& self) { return self.fd; });
+
+  auto tensor_sizes = [](nvfuser::Tensor arg) -> std::vector<nvfuser::Scalar> {
```
I'm thinking about unify tensor_sizes to a single scalar-like value.

I want to expand the concept of Scalar to a vector. something like this: https://github.com/csarofeen/pytorch/pull/2451/files#diff-df971f94c8ef070e3b8142c848e69b2d3b92054a6cf27cfc5fdc941ccf563562R76

We'll then have nested Scalar, (where a vector Scalar could hold multiple Scalar within... I know this could get messy, so I'm open to comments/suggestions as well).

jacobhinkle replied:
What would be the conceptual difference between that and a Tensor? Like I guess the nested version might support ragged multidimensional arrays?

jjsjann123 replied:
Oops, sorry about the radio silence. :stuck_out_tongue: 

It's the returned instance that I was referring to. Here we have `std::vector<nvfuser::Scalar>` as the return.
I'm thinking about changing that to a single `nvfuser::Scalar`, which will host a `std::vector<nvfuser::Scalar>`. I don't think it has any real concern for you here. Just some refactor work, so I was sort of noting this for myself.

Also, the design on that is up to discussion. Just some rough idea at this point. We can follow up when I have a PR up. :)

jacobhinkle replied:
Yeah, sounds good as long as we can still extract the "scalar" scalars.

jacobhinkle committed:
Rename to _get_fusion_definition

Also removed a junk definition of __add__

jacobhinkle committed:
Add Tensor.ndim property to frontend

This allows us to find the dimension of a tensor without creating
new Scalars, as we would if we used `x.shape` or
`fd.ops.tensor_sizes(x)`.
(Also clang-formatted code).

jacobhinkle committed:
Merge branch 'devel' into more_python_ops

jacobhinkle committed:
WIP: Add methods for most ops

I believe this commit is mostly working, but the test segfaults due to a
potential bug in addcmul. I will revisit this after that bug is fixed,
and clean up the tests then.

jacobhinkle committed:
Enable addcmul in op methods test

jacobhinkle committed:
Merge branch 'devel' into more_python_ops

jacobhinkle commented on third_party/nvfuser/csrc/python_frontend/fusion_definition.h:
```diff
@@ -192,7 +204,7 @@ class TORCH_CUDA_CU_API FusionDefinition {
     Operators(FusionDefinition* fd) : fusion_definition(fd) {}
     bool validUse() const {
-      return !fusion_definition->id().has_value();
+      return !fusion_definition->completed();
```
@kevinstephano I changed this so that `FusionDefinition` can tell us if it's completed. I needed that since I bypass `Operators` when defining methods for `Tensor` and `Scalar`, and grab the `FusionDefinition` directly from them.

jacobhinkle commented on third_party/nvfuser/csrc/python_frontend/fusion_definition.h:
```diff
@@ -59,17 +61,23 @@ struct TORCH_CUDA_CU_API Tensor {
   size_t dims;
+
+  //! Pointer to the FusionDefinition used to create this tensor
+  FusionDefinition* fd;
```
Should I rename these to `fusion_definition`?

jjsjann123 replied:
Note: this is my coding nitpick:

I'm a little bit worried about having `fd` here would make our thread safety refactor later even more complicated/slow, but that could be me being paranoid.

Are we supposed to modify the owner fusion definition through its tensor? is that also why we added the `complete` function as well? I'm wondering if this could be worked around with passing the pointer along with tensors. I have to admit that does feel cumbersome as well.

jjsjann123 replied:
OK, I see where this is coming from now, but I'm holding the same concern :)

I'm ok with us moving forward with this PR on this. cc'ing @kevinstephano for a second opinion.
Meanwhile, we should discuss the feasibility of wrapping tensor methods in python, instead of awkwardly doing it in pybind. IMHO, user experience would be better with easily accessible python scripts, vs going over pybind registration in python_bindings.cpp.

jacobhinkle replied:
I agree it'd be easier if we could implement all this in Python. See my other comment though; I just couldn't find a non-awkward way to do the wrapping and still get the expected types as return values from ops.

Holding the `fd` (will rename to `fusion_definition`) was purely a coincidence thing. If we wrap val types in Python, then we could hold it there instead.

jacobhinkle committed:
De-blackify test_python_frontend.py

I had run black on this file to format it, but it makes for a noisy PR.

jacobhinkle committed:
Merge branch 'devel' into more_python_ops

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/fusion_definition.h:
```diff
@@ -120,6 +128,10 @@ class TORCH_CUDA_CU_API FusionDefinition {
   //! Return fusion id of defined FusionDefinition
   c10::optional<size_t> id() const;
 
+  bool completed() {
```
nice~~

jjsjann123 commented on third_party/nvfuser/python_tests/test_python_frontend.py:
```diff
@@ -1065,6 +1065,137 @@ def fusion_func(fd: FusionDefinition):
+            c0 = fd.define_constant(1.0)
+
+            t1 = -t0
+            t2 = abs(t0)
```
Nice~~ Like this quite a lot.

Do you think it make sense for us to wrap these APIs in python directly, instead of going via pybind? Not suggesting that in this PR :laughing:

jacobhinkle replied:
We could, but the problem is that the ops return the C++ objects. If we subclass them, then we'd need to wrap/unwrap when calling the ops, which gets messy. That's why I went for just doing the dunders in pybind11 here.

jjsjann123 replied:
Understood. We can revisit this topic later. I'm happy with the changes from this PR on customer facing APIs.

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:
```diff
@@ -471,84 +523,179 @@ void initNvFuserPythonBindings(PyObject* module) {
-      },                                                                      \
+#define NVFUSER_PYTHON_BINDING_UNARY_OP_SPECIAL(op_str, op_name)               \
+  tensor_class.def(                                                            \
+      "__" op_str "__",                                                        \
```
oh dunder methods :clap:

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:
```diff
@@ -573,417 +720,117 @@ void initNvFuserPythonBindings(PyObject* module) {
-             fd->recordingState(arg2()),                                       \
-             fd->recordingState(arg3()),                                       \
-             fd->recordingState(arg4())},                                      \
+        FusionDefinition* fd = arg1.fd;                                        \
```
yeah, let's do `arg1.fusion_definition` instead `arg1.fd`~

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:
```diff
@@ -573,417 +720,117 @@ void initNvFuserPythonBindings(PyObject* module) {
-         Scalar arg2,                                                          \
-         Scalar arg3,                                                          \
-         Scalar arg4) -> Scalar {                                              \
+#define NVFUSER_PYTHON_BINDING_BINARY_OP_SPECIAL(py_op, op_str, op_name)       \
```
nitpick, op_str is just stringified op_name, maybe we don't need both.

jjsjann123 commented on third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:
```diff
@@ -1130,15 +1834,43 @@ void initNvFuserPythonBindings(PyObject* module) {
+            ("ops." op_str),                                                            \
+            static_cast<                                                                \
+                                                                                        \
+                TensorView* (*)(TensorView*, const std::vector<int>&, bool, DataType)>( \
```
nitpick, this is some strange formatting. clangformat is beyond my comprehension.

jacobhinkle replied:
I noticed it and have no clue why clang-format insists on that formatting.

jjsjann123 reviewed (approved):
LGTM. I'm stamping.

jacobhinkle committed:
Rename {Fusion,Scalar}.fd to .fusion_definition
